### PR TITLE
Fix node updateWorldTransform & setWorldScale

### PR DESCRIPTION
**Pull request by PatriceJiang**
Re: https://github.com/cocos/cocos-engine/issues/12861

When a parent node has non-uniform scaling, the world scale and rotation of its child nodes cannot be guaranteed to be correct at the same time.

In a non-uniformly scaled parent node, the three axes of the local coordinate system of its child nodes may no longer be perpendicular to each other. As a result, local rotation affects the scaling ratio of each axis of the child node, causing its local scale and rotation to become dependent on each other.

The new implementation ensures that the value of `getWorldScale` after `setWorldScale` on a child node is equal, even in the presence of non-uniform scaling. However, the previous implementation ensured that the value of `getWorldRotation` was equal after `setWorldRotation`, which could cause compatibility issues.

The best practice is to **avoid non-uniform scaling in parent nodes** to ensure that the local scale and local rotation of the child nodes are correct. 

### Changelog

*

-------

### Continuous Integration

This pull request:

* [x] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [x] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

**Comment by minggo** [comment]
@PatriceJiang could please describe the reason of the modification.

File: cocos/scene-graph/node.ts [hunk 1]
```diff
@@ -1986,14 +1988,9 @@ export class Node extends CCObject implements ISchedulable, CustomSerializable {
                 if (dirtyBits & TransformBit.RS) {
                     Mat4.fromRTS(child._mat, child._lrot, child._lpos, child._lscale);
                     Mat4.multiply(child._mat, cur._mat, child._mat);
-                    if (dirtyBits & TransformBit.ROTATION) {
-                        Quat.multiply(child._rot, cur._rot, child._lrot);
-                    }
```
Comment (PatriceJiang) [inline cocos/scene-graph/node.ts:1991]: `child._rot` is effected by parent's scale: `child.worldRotation != parent.worldRotation * child.localRotation`

File: cocos/core/math/mat4.ts [hunk 2]
```diff
@@ -828,30 +828,44 @@ export class Mat4 extends ValueType {
...
+            m3_1.m02 = m.m02 / sx;
+            m3_1.m03 = m.m04 / sy;
+            m3_1.m04 = m.m05 / sy;
+            m3_1.m05 = m.m06 / sy;
+            m3_1.m06 = m.m08 / sz;
+            m3_1.m07 = m.m09 / sz;
+            m3_1.m08 = m.m10 / sz;
+            const det = Mat3.determinant(m3_1);
+            if (det < 0) {
+                if (s) s.x *= -1;
```
Comment (stanleyljl) [inline cocos/core/math/mat4.ts:855]: Whether getting q or not affects the value of s.

File: cocos/core/math/mat4.ts [hunk 3]
```diff
@@ -828,30 +828,44 @@ export class Mat4 extends ValueType {
...
+        const sz = Vec3.set(v3_1, m.m08, m.m09, m.m10).length();
+        if (s) {
+            s.x = sx;
+            s.y = sy;
+            s.z = sz;
+        }
+        if (v) {
+            Vec3.set(v, m.m12, m.m13, m.m14);
+        }
+        if (q) {
```
Comment (PatriceJiang) [inline cocos/core/math/mat4.ts:843]:
```suggestion
        if (q || s) {
```

File: cocos/core/math/mat4.ts [hunk 4]
```diff
@@ -828,30 +828,44 @@ export class Mat4 extends ValueType {
...
+            m3_1.m07 = m.m09 / sz;
+            m3_1.m08 = m.m10 / sz;
+            const det = Mat3.determinant(m3_1);
+            if (det < 0) {
+                if (s) s.x *= -1;
+                m3_1.m00 *= -1;
+                m3_1.m01 *= -1;
+                m3_1.m02 *= -1;
+            }
+            Quat.fromMat3(q, m3_1); // already normalized
```
Comment (PatriceJiang) [inline cocos/core/math/mat4.ts:860]:
```suggestion
            if(q) {
                Quat.fromMat3(q, m3_1); // already normalized
            }
```

File: cocos/core/math/mat4.ts (hunk 2 above)
Comment (PatriceJiang) [reply cocos/core/math/mat4.ts:855]: fixed

**Review by stanleyljl — APPROVED** [review]

**Review by minggo — APPROVED** [review]

**Review by minggo — APPROVED** [review]

**Pull request merged**